lib/Screencast.js

```diff
@@ -0,0 +1,234 @@
+ */
+const fs = require('fs');
+const EventEmitter = require('events');
+const {helper, debugError} = require('./helper');
+
+/**
+ * @param {!Array<!Object>} frames
+ * @return {!Promise<!Array<number>>}
+ */
+async function createVideo(frames) {
+  const imageBitmaps = await Promise.all(frames.map(frame => {
+    const blob = new Blob([Uint8Array.from(frame.buffer)], {type: 'image/png'});
+    return window.createImageBitmap(blob);
+  }));
+
+  // Use last image as width so we know the set viewport is correct.
+  const WIDTH = imageBitmaps[imageBitmaps.length - 1].width;
+  const HEIGHT = imageBitmaps[imageBitmaps.length - 1].height;
+
+  const canvas = document.createElement('canvas');
+  canvas.style.display = 'none';
+  canvas.width = WIDTH;
+  canvas.height = HEIGHT;
+
+  // Create video stream from canvas. Since we're not passing an fps, it will
+  // match how fast the canvas is being drawn to.
+  const stream = canvas.captureStream();
+  const recorder = new MediaRecorder(stream, {mimeType: 'video/webm'});
+
+  let fulfill;
+  const videoPromise = new Promise(r => fulfill = r);
+
+  const setupAndStartRecording = function() {
+    return new Promise((resolve, reject) => {
+      const chunks = [];
+      recorder.ondataavailable = e => chunks.push(e.data);
+      recorder.onstop = e => resolve(chunks);
+      recorder.onerror = reject;
+      recorder.start();
+    });
+  };
+
+  const blobToArrayBuffer = function(blob) {
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.onerror = reject;
+      reader.onload = e => {
+        resolve(e.target.result);
+      };
+      reader.readAsArrayBuffer(blob);
+    });
+  };
+
+  const drawImagesToCanvas = async function() {
+    const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
+
+    const ctx = canvas.getContext('2d');
+
+    ctx.drawImage(imageBitmaps[0], 0, 0);
+    for (let i = 1; i < frames.length; ++i) {
+      const msDiff = (frames[i].metadata.timestamp - frames[i - 1].metadata.timestamp) * 1000;
+      await sleep(msDiff);
+      ctx.drawImage(imageBitmaps[i], 0, 0);
+    }
+  };
+
+  setupAndStartRecording().then(async chunks => {
+    const arrayBuffer = await blobToArrayBuffer(new Blob(chunks, {type: 'video/webm'}));
+    // Convert to array for serialization back to pptr.
+    fulfill(Array.from(new Uint8Array(arrayBuffer)));
+  });
+
+  await drawImagesToCanvas(); // Draw images to canvas in real-time.
+
+  recorder.stop();
+
+  return videoPromise;
+}
+
+class Screencast extends EventEmitter {
+  /**
+   * @param {!Session} client
+   */
+  constructor(client) {
+    super();
+    this._client = client;
+    /** @type {!Array<!Object>} */
+    this._frames = [];
+    /** @type {string} */
+    this._format = 'png';
+    /** @type {string} */
+    this._path = '';
+    /** @type {boolean} */
+    this._recording = false;
+
+    this._client.on('Page.screencastFrame', event => this._onScreencastFrame(event));
```

Comment:
we don't want to use this api for video streaming; it gonna be slow and low-quality. We'd need to introduce a nicer API for the protocol first to have this feature in pptr.